python_submitty_utils/submitty_utils/submitty_schema_validator.py

```diff
@@ -0,0 +1,186 @@
+                validate_schema(v, spec_v_schema, prefix=t_name, warn=warn)
+            except SubmittySchemaException:
+                raise
+            validator_num += 1
+        testcase_num += 1
+
+        try:
+            validate_schema(t, t_s, prefix=t_name, warn=warn)
+        except SubmittySchemaException:
+            raise
+    # Finally, validate the config as a whole.
+    try:
+        validate_schema(j_, s_, prefix='Your config json', warn=warn)
+    except SubmittySchemaException:
+        raise
+
+
+def validate_schema(j_, s_, key='', prefix='', required=True, warn=False):
+    """
+    Validate a configuration against a schema.
+
+    A function which validates a key within a config and a schema.
+    If no key is given, we evaluate the whole config or schema.
+    """
+    descriptive_title = '{0} {1}'.format(prefix, key)
+    if key != '':
+        my_json_chunk = j_.get(key, None) if j_ is not None else j_
+        my_schema = s_.get(key, None) if s_ is not None else s_
+    else:
+        my_json_chunk = j_
+        my_schema = s_
+    # If attempting to validate an item not specified in the schema, fail.
+    if my_schema is None:
+        raise SubmittySchemaException(j_,
+                                      s_,
+                                      ("ERROR: There is no specification for "
+                                       "{0} in the schema. Please "
+                                       "add a specification.").format(key),
+                                      descriptive_title,
+                                      None)
+    if my_json_chunk is None:
+        if warn:
+            print(("WARNING: could not identify "
+                   "{0} ({1})").format(descriptive_title, key))
+        return
+    try:
+        # validate.
+        validate(instance=my_json_chunk, schema=my_schema)
+    except Exception as e:
+        raise SubmittySchemaException(my_json_chunk,
+                                      my_schema,
+                                      ('ERROR: {0} was not properly '
+                                       'formatted')
+                                      .format(descriptive_title),
+                                      descriptive_title,
+                                      e)
+
+
+def validate_complete_config_schema_using_filenames(config_path,
+                                                    schema_path,
+                                                    show_warnings=True):
+    """
+    Call validate_config after reading in config and schema files.
+
+    Given a path to a complete configuration and a schema, this function
+    loads both and validates the configuration against the schema. On
+    failure, an exception is thrown, else, nothing is returned.
+    """
+    with open(schema_path, 'r') as infile:
+        schema = jsonref.load(infile)
+
+    with open(config_path, 'r') as infile:
+        config_json = jsonref.load(infile)
```

Comment:
Are we allowing the config jsons to have references or must they be standard json?

Reply:
Good catch. At the moment, configs must be standard JSON. Schemas can have references, however, as that makes them far more writable.

Reply:
I'm very much against supporting jsonref in regular configs as it's not a standard feature and that there's other formats that already support that and are better suited for what we're doing, such as YAML.